students/kristianjf/lesson10/mailroom_6.py

```diff
@@ -0,0 +1,302 @@
+        through all the donors in your donor data structure, generates a thank you \
+        letter, and writes it to disk as a text file.
+        '''
+        filename = '_'.join(self.name.split())+'.txt'
+        with open(filename, 'w') as outfile:
+            outfile.write(self.output_string().format(name=self.name, donation=self.donations))
+            print(f'Writing: {filename}')
+
+    def output_string(self):
+        'Output string for letters'
+        format_string = 'Dear {name},\n\nThank you for your generous donation of ' \
+                        '${donation}. Please send us more money at your earliest ' \
+                        'convenience.'
+        return format_string
+
+class DonorHandler:
+    def __init__(self):
+        self.donors = {}
+    def add_donor(self, donor_object):
+        self.donors[donor_object.normalized_name] = donor_object
+    def dnr_challenge(self, factor=1, **kwargs):
+        new_dh = DonorHandler()
+        for donor, donor_object in self.donors.items():
+            # new_d = Donor(donor, list(map(lambda x, y: x*y, \
+            # donor_object.donations, len(donor_object.donations)*[factor])))
+            # new_d = Donor(donor, \
+            # self.dnr_filter(\
+            # list(map(lambda x, y: x*y, donor_object.donations, \
+            # len(donor_object.donations)*[factor])),\
+            # kwargs['min_donation'], kwargs['max_donation']))
+            # new_dh.add_donor(new_d)
+            '''Add a new feature to Mailroom using filter so that donations either above
+            or below a specified dollar amount are included in the map operations of #1 above.
+            You can do this by adding min_donation and max_donation optional keyword parameters
+            to your challenge function. You’ll want to filter the donations before
+            passing them to map.'''
+            filter_donations = self.dnr_filter(\
+            donor_object.donations, kwargs['min_donation'], kwargs['max_donation'])
+
+            '''Add a new feature to Mailroom using map so that each donation on record can be
+            doubled, tripled or indeed multiplied by any arbitrary factor based on the whims
+            of philanthropists who would like to support our cause.
+            This will require a new function (or method in your donor database class) called
+            challenge(factor) that takes a multiplier (factor), and multiplies all the
+            donations of all the donors by the factor. The function returns a NEW donor database,
+            with the new data.'''
+            map_donations_f = list(map(lambda x, y: x*y, filter_donations,\
+            len(filter_donations)*[factor]))
+
+            new_d = Donor(donor, map_donations_f)
+            new_dh.add_donor(new_d)
+        return new_dh
+    def dnr_filter(self, lst, min_donation=0, max_donation=10000000):
+        lst = list(filter(lambda x: (x > min_donation) and (x < max_donation), lst))
```

Comment:
do you need to make a list? if it's going to be looped over, it could stay an iterable.
that's why map, filter, zip, etc return iterables -- so they can be looped over without creating a whole new sequence.